README.md
# beautifulplots

Python plotting library for beautiful, easy, and sophisticated matplotlib based plots. Compatible with Matplotlib graphs, such as Pandas and Seaborn. 

Python, along with Pandas, and Nummpy are the data scientist analysis tool of choice. Correspondingly, data visualization is required for analysis, investigation, and insight. In this endeavor, Matplotlib based plotting libraries are reputedly the most powerful and customizable. However, creating the necessary analysis graphs, though necessary, often diverges from the primary analysis goal. Graphs resulting from data science analysis inevitably populate reports, presentations, and even interactive web graphics. After some experience, use case after use case requires very similar surrounding code. 

The end goal of the beautifulplots plotting library is to package graphing functions for the easy generation of beautiful graphs without the burden of distracting code. 

## Installation

```bash
$ pip install beautifulplots
```

## Usage

- See usage examples in examples.ipynb (Jupyter notebook) and readthedocs [beatifulplots readthedocs](https://beautifulplots.readthedocs.io/en/latest/index.html).


## Feature Requests and Issues

Feature requests and issues: please provide feedback and suggestions for improvements or issues via github issues [new issue](https://github.com/Aljgutier/beautifulplots/issues).


## License

`beautifulplots` was created by Alberto Gutierrez. It is licensed under the terms of the MIT license.

## Credits

`beautifulplots` package setup with  [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.1.0 (29/06/2022)

- First release of `beautifulplots`!

tests/test_beautifulplots.py
import beautifulplots as bp
import pandas as pd


def test_plot_defaults():
    """Test plot defaults."""

    
    plot_defaults_expected={'pltstyle': 'seaborn','figsize': None,'legend_loc': 'best',
    'legend_loc2': 'best','wspace': 0.2,'hspace': 0.2,'sharex': False,'title': None,
    'ylims': None, 'xlims': None, 'ylims2': None,
    'xlabelfontsize': 16,   'xticklabelsize': 16, 'xtickfontsize': 16,
    'xtickrotation': 0, 'ylabelfontsize': 16, 'ytickfontsize': 16, 'ytickrotation': 0,
    'titlefontsize': 18, 'legendsize': 16, 'legend': False, 'xlabel': '', 'ylabel': '',
    'color': None, 'color2': None, 'palette': None, 'palette2': None, 'hue': None,
    'ci': None, 'ci2': None, 'marker': None, 'marker2': None, 'yaxis_currency': False,
    'ytick_format': None, 'legend_labels': None, 'alpha': None, 'alpha2': 0.5, 
    'estimator': sum, 'estimator2': sum}
    
    plot_defaults_actual = bp.plot_defaults()
    assert plot_defaults_actual == plot_defaults_expected, "plot_defaults incorrect"
    
    
def test_barplot():
    
    # data 
    print('Data:Product Category Sales by Store')
    barplot_data = { 'category':['groceries','groceries','groceries', 'hardware','hardware','hardware','hobbies','hobbies','hobbies'] ,
                'sales':[ 90, 60,50, 50,30,20,40,40,20], 'store':['A','B','C','A','B','C','A','B','C'] }

    df = pd.DataFrame(barplot_data)
    # create unit sales by category ... assume some average sale price per category
    def unit_sales(row):
        units = 0
        if row['category'] == 'groceries': units = row['sales']/5
        elif row['category'] == 'hardware': units = row['sales']/10
        elif row['category'] == 'hobbies': units = row['sales']/20
        return units
    df['units']= df.apply(lambda row: unit_sales(row),axis=1)
    
    
    # test that barplot completes normally and returns None
    plot_return = bp.barplot(df,'category','sales',hue='store', palette='tab20', 
                   title='1e. easy barplot, Product Category Sales by Store, horizontal bars', figsize=(12,8), 
                   bardatalabels=True, bardataformat=".2f", barcurrency=True, bar_orientation='h',
                   xlims = (0.1,100), legend=True, legendloc="lower right", test_mode=True)
    assert plot_return == None, "barplot error: did not complete normally"
    
    # do not set bardataformat ... ensure it has useful default
    plot_return = bp.barplot(df,'category','sales',hue='store', palette='tab20', 
                   title='1e. easy barplot, Product Category Sales by Store, horizontal bars', figsize=(12,8), 
                   bardatalabels=True, barcurrency=True, bar_orientation='h',
                   xlims = (0.1,100), legend=True, legendloc="lower right", test_mode=True)
    
    assert plot_return == None, "barplot error: did not complete normally"
    
    # do not set bardataformat ... use color
    plot_return = bp.barplot(df,'category','sales',hue='store', color='blue', 
                   title='1e. easy barplot, Product Category Sales by Store, horizontal bars', figsize=(12,8), 
                   bardatalabels=True, barcurrency=True, bar_orientation='h',
                   xlims = (0.1,100), legend=True, legendloc="lower right", test_mode=True)
    assert plot_return == None, "barplot error: did not complete normally"
    
    # no color or palette, vertical bars
    # do not set bardataformat ... no color or palette
    plot_return = bp.barplot(df,'category','sales',hue='store', 
                   title='1e. easy barplot, Product Category Sales by Store, horizontal bars', figsize=(12,8), 
                   bardatalabels=True, bar_orientation='v',
                   xlims = (0.1,100), legend=True, legendloc="lower right", test_mode=True)
    assert plot_return == None, "barplot error: did not complete normally"
    
    
    bp.barplot(df,'category','sales', palette='tab20', 
                   title='1a. Beautifulplots barplot, Sales All Stores by Product Category, secondary y-axis', figsize=(12,4),
                   bardatalabels=True, bardataformat=".2f", barcurrency=True, ylims = (0.1,220), 
                   y2='units',ylims2=(0,45), color2='black',marker2="o", test_mode=True)
    assert plot_return == None, "barplot error: did not complete normally"
    
def test_lineplot():
    
    # data
    sp500_file =  './data/GSPC_1950-1-3_to_2022-6-8.csv'
    df_sp500 = pd.read_csv(sp500_file,index_col=0,parse_dates=True)
    # reset index ...  Seaborn requires x-axis to be a column 
    df = df_sp500.reset_index() 

    
    plot_return=bp.lineplot(df,x='Date' , y=['Close','Open'], y2='Volume', 
             ylims=(0,4000), ylims2=(0,10*1e9), legend_loc2 = "lower left", color2='red', alpha2=0.5,
             figsize=[12,4],yaxis_currency=True, legend=True,
             ytick_format=".0f", title="S&P 500, easy lineplot", test_mode=True)
    
    assert plot_return == None, "lineplot error: did not complete normally"
    
    # palette 
    plot_return=bp.lineplot(df,x='Date' , y=['Close','Open'], y2='Volume', palette='tab20',
             ylims=(0,4000), ylims2=(0,10*1e9), legend_loc2 = "lower left", color2='red', alpha2=0.5,
             figsize=[12,4],yaxis_currency=True, legend=True,
             ytick_format=".0f", title="S&P 500, easy lineplot", test_mode=True)
    
    assert plot_return == None, "lineplot error: did not complete normally"
    
    
    plot_return=bp.lineplot(df,x='Date' , y=['Close','Open'], y2='Volume', color='red',
             ylims=(0,4000), ylims2=(0,10*1e9), legend_loc2 = "lower left", color2='red', alpha2=0.5,
             figsize=[12,4],yaxis_currency=True, legend=True,
             ytick_format=".0f", title="S&P 500, easy lineplot", test_mode=True)
    
    assert plot_return == None, "lineplot error: did not complete normally"

src/beautifulplots/barplot.py

import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib.colors import Normalize
from matplotlib.cbook import boxplot_stats
import matplotlib as mpl

import  beautifulplots.beautifulplots as bp  


def barplot(df, bar_columns, bar_values, barcurrency=None, barorientation="v", bardataformat=None,
            y2=None, ax=None, bardatalabels=False, test_mode=False, bardatafontsize=14,
            **kwargs):
    """Bar plot function designed for ease of use and aesthetics. 
    The underlying barplot is ased on the Seaborn with additions, such as secondary axis, data labels,
    and improved default parameters. Refer to beautifulplots plot_defaults for a complete list of options.
    
    Args:
        df (DataFrame): The input DataFrame containing colums corresponding to bar_plot values ("bar_values") and column names (see examples in documentation)
            
        bar_columns: Datafrae columns corresponding to bar column names
            
        bar_values: Dataframe column corresponding to bar column values
            
        ax (axis): matplotlib axis (optional), default = None. If axis is None, then create a matplolib figure, axis to host the barplot
            
        color: Matplotlib compatabile color name as text or RGB values, for example,  color = [51/235,125/235,183/235].
            
        palette: Matplotlib compatible color palette name, for example, "tab20"
            
        hue: Name of hue dimension variable (i.e., DataFrame column name)
            
        ci: Seaborn confidence interval parameter: float, sd, or None, default = None
            
        barorientation: default = v (vertical), or h (horizontal)
            
        barcurrency: default = False (bar values do not represent currency). True (bar values represent currency, append $ to the value)
            
        bardatalabels (Boolean): default = False (data labels not included)
        
        additional options:  see kale.plot_defaults for additional input variables.
            

    Returns:
        returns True if processing completes succesfully (without errors).
    """
        
    plot_options = bp.get_kwargs(**kwargs)
    
    estimator = plot_options['estimator'] 
    estimator2 = plot_options['estimator'] 
    ci = plot_options['ci']
    ci2 = plot_options['ci2']
    alpha = plot_options['alpha']
    alpha2 = plot_options['alpha2']
    hue = plot_options['hue']
    palette = plot_options['palette']
    palette2 = plot_options['palette2']
    marker2 = plot_options['marker2']
    color = plot_options['color']
    color2 = plot_options['color2']


    # if no hue then only one color
    # if hue == None and color==None : color = [51/235,125/235,183/235] if plot_options['color'] == None else plot_options['color']

    if barorientation == 'v': x,y = bar_columns, bar_values
    else: x,y = bar_values, bar_columns
    

    if ax == None: 
        mpl.rcParams.update(mpl.rcParamsDefault) # reset plot/figure parameters
        plt.style.use(plot_options['pltstyle'])
        fig,_ax = plt.subplots(nrows=1, ncols=1, figsize=plot_options['figsize']) 
    else: _ax = ax
        
    g=sns.barplot(x=x, y=y, hue=hue, color = color, palette=palette, data=df, ax = _ax,
                  orient=barorientation, ci=ci, estimator=estimator, alpha=alpha)
    
    # Bar labels ... iterate with hue
    # Matplotlib
      # https://matplotlib.org/stable/gallery/lines_bars_and_markers/bar_label_demo.html#sphx-glr-gallery-lines-bars-and-markers-bar-label-demo-py
    # Geeks for Geeks
      # https://www.geeksforgeeks.org/how-to-show-values-on-seaborn-barplot/
    if bardatalabels== True and hue == None:
        f = "%"+bardataformat+""
        if barcurrency == True: f = "$"+f # dollar
        g.bar_label(g.containers[0],  fontsize=bardatafontsize, fmt=f)

    # Geeks for Geeks bar data labels
      # https://www.geeksforgeeks.org/how-to-show-values-on-seaborn-barplot/
    if  bardatalabels == True and hue !=None:
        f = "%"+bardataformat+""
        if barcurrency == True: f = "$"+f # dollar
        for i in g.containers:
            g.bar_label(i,fontsize=bardatafontsize, fmt=f )
   
    # yaxis tick label format
    # https://matplotlib.org/stable/gallery/pyplots/dollar_ticks.html
    # x or y format same as bars ... since this could be v or h graph
    if barcurrency == True:
        #y_ticks = _ax.get_yticks()
        f='{x:'+ bardataformat  +'}'
        if barcurrency== True: f='${x:'+ bardataformat  +'}'
        if barorientation=='v':
            _ax.yaxis.set_major_formatter(f)
        if barorientation=='h':
            _ax.xaxis.set_major_formatter(f)
 
   
   # secondary y-axis
    if y2 != None:
       
        if isinstance(y2,list):
            y2_list = y2
        else:
            y2_list = [y2]

        _ax2 = _ax.twinx()
        
        
        for _y2 in y2_list:
            if plot_options['palette2'] !=None:
                g = sns.lineplot(data=df,x=x, y =_y2, hue=hue, palette=palette2,  ax=_ax2, label=_y2, 
                                 alpha = alpha2,ci = ci2, marker=marker2, estimator=estimator2)
            elif plot_options['color2'] !=None:
                g = sns.lineplot(data=df,x=x, y=_y2, hue=hue, color=color2,  ax=_ax2,label=_y2, 
                                 alpha=alpha2, ci=ci2, marker=marker2, estimator=estimator2)
            else:
                g = sns.lineplot(data=df,x=x, y=_y2, hue=hue, ax=_ax2, label=_y2, 
                                 alpha=alpha2, ci=ci2, marker=marker2, estimator=estimator2) 
                
        _ax2.grid(b=None)  
    
    # set axis params
    bp.set_axisparams(plot_options,_ax,g)  # axis parameters from the plot_options dictionary
    
    # y2 axis params
    if y2 != None:
        bp.set_axisparams(plot_options,_ax2,g)  # axis parameters
    
        # set ylims 2 after general axis parameters 
        if plot_options['ylims2'] != None:
            _ax2.set_ylim(plot_options['ylims2'])
        
        # axis 2 legend
        if y2 != None and plot_options['legend']==True:
            _ax2.legend( loc=plot_options['legend_loc2'], prop={'size': plot_options['legendsize']})
    

    if ax==None and test_mode==False: plt.show() # if barplot created the figure then plt.show()
    
    return None

src/beautifulplots/lineplot.py
import matplotlib.pyplot as plt
import seaborn as sns
import matplotlib as mpl

import  beautifulplots.beautifulplots as bp 

def lineplot(df, x, y, y2=None, ax=None, test_mode=False,  **kwargs):
    """Lineplot function designed for ease of use and aesthetics. Based on the
    Seaborn lineplot function with improvements, such as secondary axis, ease of use, and 
    improved default parameters. Refer to beautiful plot_defauts for full list of options.
 
    Args:
        df (Dataframe): The input DataFrame containing colums corresponding to x and y     
            
        x: Dataframe column corresponding to the lineplot x-axis
            aldfsd;lfj
        
        y: Dataframe column corresponding to the lineplot y-axis
            
        y2: Column name correspondng to the secondary axis, default = None 

    Returns:
        returns None if processing completes succesfully (without errors).
    """ 
    plot_options = bp.get_kwargs(**kwargs)
    estimator = plot_options['estimator']
    estimator2 = plot_options['estimator2']
    hue = plot_options['hue']
    color = plot_options['color']
    ci2 = plot_options['ci2']
    alpha = plot_options['alpha']
    alpha2 = plot_options['alpha2']
    hue = plot_options['hue']
    palette = plot_options['palette']
    palette2 = plot_options['palette2']
    marker = plot_options['marker']
    marker2 = plot_options['marker2']
    color = plot_options['color']
    color2 = plot_options['color2']
    
    # get back to the default plot options
    if ax == None: 
        mpl.rcParams.update(mpl.rcParamsDefault) # reset plot/figure parameters
        plt.style.use(plot_options['pltstyle'])
        fig,_ax = plt.subplots(nrows=1, ncols=1, figsize=plot_options['figsize']) 
    else: _ax = ax
    

    # Primary Y ... loop over each element
    if isinstance(y,list):
        y_list = y
    else:
        y_list =[y]
    for _y in y_list:
        if plot_options['palette'] !=None:
            g = sns.lineplot(data=df,x=x, y=_y, hue=hue, palette=palette,  ax=_ax, label=_y, 
                             alpha = alpha, ci=ci2, marker=marker, estimator=estimator)
        elif plot_options['color'] !=None:
            g = sns.lineplot(data=df,x=x,y=_y, hue=hue, color=color,  ax=_ax,label=_y, 
                             alpha = alpha, markers=marker, estimator=estimator)
        else:
            g = sns.lineplot(data=df,x=x,y=_y, hue=hue, ax=_ax, label=_y,
                             alpha= alpha, markers=marker, estimator=estimator  )
            
            
    # second y_axis ... plot this first so that primary y is plotted over secondary
    if y2 != None:
        if isinstance(y2,list):
            y2_list = y2
        else:
            y2_list = [y2]
    
        _ax2 = _ax.twinx()
        
        for _y2 in y2_list:
            if plot_options['palette2'] !=None:
                g = sns.lineplot(data=df,x=x, y =_y2, hue=hue, palette=palette2,  ci=ci2,
                                 ax=_ax2, label=_y2, alpha = alpha2, markers=marker2,
                                 estimator=estimator2)
            elif plot_options['color2'] !=None:
                g = sns.lineplot(data=df,x=x, y=_y2, hue=hue, color=color2,  ci=ci2,
                                 ax=_ax2,label=_y2, alpha=alpha2, marker=marker2,
                                 estimator=estimator2)
            else:
                g = sns.lineplot(data=df,x=x, y=_y2, hue=hue, ax=_ax2, label=_y2, ci=ci2,
                                 alpha=alpha2, marker=marker2,
                                 estimator=estimator2) 
                
        _ax2.grid(b=None)        
            
    # yaxis currency
    if plot_options["yaxis_currency"] == True:
        if plot_options['ytick_format'] !=None:
            f="${x:"+plot_options["ytick_format"]+"}"
            _ax.yaxis.set_major_formatter(f)
        else:
            _ax.yaxis.set_major_formatter('${x:.2f}')
           
    # legend
    if plot_options['legend_labels']!=None:
        _ax.legend(labels=plot_options['legend_labels'])
    
    if plot_options['legend']==True:
        _ax.legend( loc=plot_options['legend_loc'], prop={'size': plot_options['legendsize']})  
    
    # Set Axis Parameters
    bp.set_axisparams(plot_options,_ax,g)  # axis parameters
    if y2 != None:
        bp.set_axisparams(plot_options,_ax2,g)  # axis 2 parameters
        
        # set ylims after set_axis ... set_axis lims defaults to primary y axis 
        if plot_options['ylims2'] != None:
            _ax2.set_ylim(plot_options['ylims2'])
        
        # axis 2 legend
        if y2 != None and plot_options['legend']==True:
            _ax2.legend( loc=plot_options['legend_loc2'], prop={'size': plot_options['legendsize']})
        
    # plot show if easy lineplot created the figure
    if ax==None and test_mode==False: plt.show() # if simpl_plot created the figure then plt.show()
    
    return None

src/beautifulplots/beautifulplots.py

import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib.colors import Normalize
from matplotlib.cbook import boxplot_stats
import matplotlib as mpl 


def plot_defaults():
    """Dictionary of plot parameters. Each parameter coresponds and corresponding value. 
    See also get_plot_options for extracting plot options from **kwargs. 

    **Axis - x, y, and plot area parameters**
    
    Args:
        df (DataFrame): The input DataFrame containing colums corresponding to bar values and columns.
 
        title (String): corresponds to the axis title. default = ''
            
        titlefontsize: font size of the axis title, default = 18
            
        legend_loc(String): Matplotlib legend location, for example, upper right , default = "best".
        
        legend_loc2 (String): Secondary axis legend location, for example, upper right , default = best.
         
        xlims: (xmin, xmax), minimum and maximum x-values of the axis. default = None, in which case the min and max are set automatically by matplotlib.
        
        ylims: (ymin,ymax), minimum and maximum y-values of the axis. default = None, in which case the min and max are set automatically by matplotlib.
        
        ylims2: (ymin,ymax), minimum and maximum y-values of the secondary axis. default = None, in which case the min and max are set automatically by matplotlib.
        
        xlabelfontsize: default = 16
        
        xticklabelsize: default = 16
        
        xtickfontsize: default = 16
        
        xtickrotation: default = 0
        
        ylabelfontsize: ylabel font size, default = 16
            
        ytickfontsize: xtick label font size, default = 16
        
        ytickrotation:  Rotation of the xtick label, default = 0
        
        legendsize: legend font size, default = 16
        
        legend (Boolean):  Place a legend on the axis, default = False
            
        xlabel (String): xlabel title, default = ''
            
        ylabel (String):  ylable title, default = ''
        
    **Bar, Line, Scatter Plots**
    
    Args:
        
        marker: Matplotlib line markers. default = None (Matplotlib default).
            
        marker2: Secondary axis, Matplotlib line marker. default = None (Matplotlib default).
            
        yaxis_currency (Boolean): Boolean default = False.
            
        ytick_format (String): String default = None (Matplotlib default).
            
        alpha (fraction): Trasnparancy (opacity), default = None (not transparent)
            
        alpha2 (fraction): Secondary axis, default = 0.5, 50% opacity
        
        legend_labels (list): Overide default legend labels. default = None (do not override)
        
        estimator: seaborn barplot summary estimator, default = sum
            
        estimator2: secondary axis, seaborn barplot summary estimator, default = sum
            
        color: default = None, indicateing Matplolib default (Matplotlib default)
            
        color2: secondary axis line, bar or color. defualt = None (Matplotlib default)
            
        palette: colormap, default = None
            
        palette2: colormap, default = None, secondary y axis.
            
        hue: dimension value for corresponding Seaborn graphs, default = None.
        
        ci: Seaborn confidence interval parameter: float, sd, or None
            
        ci2: Seaborn confidence interval parameter second y axis: float, sd, or None
     
    **Plots and subplots**
    
    Parameters corresponding to the plot or subplot characteristics. They are used when matplot_helpers
    functions create the plot figure and axis, otherewise, these parameters do not affect the plot.
    
    Args:
    
        plotstyle (String): matplotlib plotstyle
            
        figsize: total size (height, width) in inches of the figure, including total plotting area of all subplots and spacing
        
        wspace: width space (horizontal) between subplots, default wspace = 0.2
        
        hspace: height space (vertical) between subplots, default hspace = 0.2
            
        legendloc: default=best
   
    **Returns**
    
        Dictionary: {parameter1:value1, parameter2:value2, ... }. 
        Pairs of plat parameters and corresponding values.
    """    

    plot_defaults = {
    # plot and subplot
    'pltstyle': 'seaborn',
    'figsize' :  None,
    'legend_loc':'best',
    'legend_loc2':'best',
    'wspace': 0.2,
    'hspace': 0.2,
    'sharex':False,
    
    # Axis ( corx and y plot)
    'title' :None,               #  list ot titles, 1 per axis  #  list of x,y tuples
    'ylims' :None,               # list of two-tuples ylims (lower, upper)
    'xlims' :None,
    'ylims2':None,
    'xlabelfontsize':16,
    'xticklabelsize':16,
    'xtickfontsize': 16,
    'xtickrotation':0,
    'ylabelfontsize':16,
    'ytickfontsize':16,
    'ytickrotation':0,
    'titlefontsize':18,
    'legendsize':16,
    'legend':False,
    'xlabel': '',
    'ylabel': '',
    'color':None,   # color designation for the corresponding graph
    'color2':None,   # color secondary y axis
    'palette':None,  
    'palette2':None,  # palette secondary y axis
    'hue':None,   # dimensional value for corresponding Seaborn graphs
    'ci':None, # confidence parameter for pimary axis
    'ci2': None, # confidence parameter for secondary axis


    # Lines and Scatter
    'marker':None,  # Boolean, list, dictionary
    'marker2':None,  # Boolean, list, or dictionary
    'yaxis_currency':False,
    'ytick_format': None,
    'legend_labels':None,
    'alpha': None,
    'alpha2': 0.5,
    'estimator':sum,
    'estimator2':sum
 

    }

    return plot_defaults

def set_axisparams(options_dict,ax,g): 
    """Receives as input a dictionary of plot options and applies the options to the maxtplotlib axis and graph.

    Args:
        options_dict (dictionary): Dictionary where each key, value pair corresponds to plot parameter
        ax (axis): matplotlib axis to apply the plot options
        g (graph): matplotlib graph to apply the plot options

    Returns:
        None: returns None if the function completes without errors.
    """ 

    title=options_dict['title']
    titlefontsize=options_dict['titlefontsize']
    legendloc=options_dict['legend_loc']
    legendsize=options_dict['legendsize']
    xlabel=options_dict['xlabel']
    xlabelfontsize=options_dict['xlabelfontsize']
    xlims=options_dict['xlims']
    xtickfontsize=options_dict['xtickfontsize']
    xtickrotation=options_dict['xtickrotation']
    ylabel=options_dict['ylabel']
    ylabelfontsize=options_dict['ylabelfontsize']
    ytickfontsize=options_dict['ytickfontsize']
    ytickrotation=options_dict['ytickrotation']
    ylims=options_dict['ylims']
    
    
    ax.set_xlabel(xlabel,fontsize=xlabelfontsize)
    ax.set_ylabel(ylabel,fontsize=ylabelfontsize)

    for x_tick in ax.get_xticklabels():
        x_tick.set_fontsize(xtickfontsize)
        x_tick.set_rotation(xtickrotation)

    for y_tick in ax.get_yticklabels():
        y_tick.set_fontsize(ytickfontsize)
        y_tick.set_rotation(ytickrotation)


    ax.set_title(title, fontsize=titlefontsize)

    if ylims != None:
        ax.set_ylim(ylims[0],ylims[1])

    if xlims != None:
        ax.set_xlim(xlims[0],xlims[1])

    if options_dict['legend'] == True:
        ax.legend( loc=legendloc, prop={'size': legendsize})     
        

    return None

def get_kwargs(**kwargs):   
    """process **kwargs options corresponding to the plot_defaults dictionary (see above)
    If a beautifulplots plot_option dictionary key is 
    contained in the **kwargs then the plot_defautls[key] value is replaced 
    with that found in **kwargs. 

    Returns:
        Dictionary: {parameter1:value1, parameter2:value2, ...} dictionary corresponding to plot options 
    """   
       
    plot_options = plot_defaults() # returns a dictionary of defaut matplotlib parameters

    # interate through the parameters in the plot_options dictionary
    # find the key in the kwargs, otherwise the paremeer = default
    # if the kwargs parameter != default then update plot_options
    for key in plot_options:
        default = plot_options[key]
        kwarg_value = kwargs.get(key, default) # parameter = kwarg  or default
        if kwarg_value != default: plot_options[key] = kwarg_value # update plot_option 
        
    return plot_options